cmd/bootstrap/run/epochs.go

```diff
@@ -0,0 +1,200 @@
+	collectors := currentEpochIdentities.Filter(filter.HasRole[flow.Identity](flow.RoleCollection))
+	internalCollectors := make(flow.IdentityList, 0)
+	partnerCollectors := make(flow.IdentityList, 0)
+
+	log.Info().Msg("collecting internal node network and staking keys")
+	internalNodes, err := common.ReadFullInternalNodeInfos(log, internalNodePrivInfoDir, nodeConfigJson)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read full internal node infos: %w", err)
+	}
+
+	internalNodesMap := make(map[flow.Identifier]struct{})
+	for _, node := range internalNodes {
+		if !currentEpochIdentities.Exists(node.Identity()) {
+			return nil, fmt.Errorf("node ID found in internal node infos missing from protocol snapshot identities %s: %w", node.NodeID, err)
+		}
+		internalNodesMap[node.NodeID] = struct{}{}
+	}
+	log.Info().Msg("")
+
+	for _, collector := range collectors {
+		if _, ok := internalNodesMap[collector.NodeID]; ok {
+			internalCollectors = append(internalCollectors, collector)
+		} else {
+			partnerCollectors = append(partnerCollectors, collector)
+		}
+	}
+
+	currentEpochDKG, err := epoch.DKG()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get DKG for current epoch: %w", err)
+	}
+
+	log.Info().Msg("computing collection node clusters")
+
+	assignments, clusters, err := common.ConstructClusterAssignment(log, partnerCollectors, internalCollectors, collectionClusters)
+	if err != nil {
+		log.Fatal().Err(err).Msg("unable to generate cluster assignment")
+	}
+	log.Info().Msg("")
+
+	log.Info().Msg("constructing root blocks for collection node clusters")
+	clusterBlocks := GenerateRootClusterBlocks(epochCounter, clusters)
+	log.Info().Msg("")
+
+	log.Info().Msg("constructing root QCs for collection node clusters")
+	clusterQCs := ConstructRootQCsForClusters(log, clusters, internalNodes, clusterBlocks)
+	log.Info().Msg("")
+
+	dkgPubKeys := make([]cadence.Value, 0)
+	nodeIds := make([]cadence.Value, 0)
+
+	// NOTE: The RecoveryEpoch will re-use the last successful DKG output. This means that the consensus
+	// committee in the RecoveryEpoch must be identical to the committee which participated in that DKG.
+	dkgGroupKeyCdc, cdcErr := cadence.NewString(currentEpochDKG.GroupKey().String())
+	if cdcErr != nil {
+		log.Fatal().Err(cdcErr).Msg("failed to get dkg group key cadence string")
+	}
+	dkgPubKeys = append(dkgPubKeys, dkgGroupKeyCdc)
+	for _, id := range currentEpochIdentities {
+		if id.GetRole() == flow.RoleConsensus {
+			dkgPubKey, keyShareErr := currentEpochDKG.KeyShare(id.GetNodeID())
+			if keyShareErr != nil {
+				log.Fatal().Err(keyShareErr).Msg(fmt.Sprintf("failed to get dkg pub key share for node: %s", id.GetNodeID()))
+			}
+			dkgPubKeyCdc, cdcErr := cadence.NewString(dkgPubKey.String())
+			if cdcErr != nil {
+				log.Fatal().Err(cdcErr).Msg(fmt.Sprintf("failed to get dkg pub key cadence string for node: %s", id.GetNodeID()))
+			}
+			dkgPubKeys = append(dkgPubKeys, dkgPubKeyCdc)
+		}
```

Comment:
> For recovery epochs, we might need to be able to include consensus nodes that are ejected

This is my understanding -- we do need to include consensus nodes that are ejected. The committee and indices need to be exactly the same.

Reply:
that is also my understanding. Though I still have the vague hope that this might not be absolutely necessary. Trying to set up a conversation with Tarak...

Reply:
Copying some context here:
[Slack thread](https://flow-foundation.slack.com/archives/C071612SJJE/p1720629919474609)

> https://github.com/onflow/flow-go/blob/master/module/signature/threshold.go#L7
here we need to pass in the size of the original DKG committee (qhich equals the length of the []PublicKey shares vector)

> For this function (it is the recipient of the DKG information):
https://github.com/onflow/crypto/blob/main/bls_thresholdsign.go#L113
we concluded that ideally sharePublicKeys []PublicKey would be a map from DKG's threshold participant index to PublicKey.
However, we thought that this change is not absolutely necessary, if we relied on the requirement that DKG indices are consecutive from 0,...,n-1 for n the number of DKG participants.

> In summary, I think there are two conventions currently used in the code base:
convention that the DKG committee is identical to the consensus committee
this is the convention we want to generalize.
Specifically, we want to allow that the consensus committee is a subset of the DKG committee (this is important when we have ejected nodes, that participated in the DKG but since they are ejected are not part of the consensus committee for the recovery epoch). Lets refer to the set of nodeIDs that participated in the DKG as D
furthermore, we want to future-proof our implementation and allow the consensus committee to be larger than the DKG committee (because consensus can likely scale to a larger number of participants than DKG can). Lets refer to the set of nodeIDs in the consensus committee as C
therefore, we conclude (Yurii phrased this very nicely): the mapping from nodeID to DKG-index would have the the set intersection D ∩ C as its [domain](https://en.wikipedia.org/wiki/Domain_of_a_function); the set of DKG-indices would be the map's [codomain](https://en.wikipedia.org/wiki/Codomain).
An independent degree of freedom in my opinion is the convention whether the set of DKG-indices is always a consecutive sequence 0,...,n-1 for n the number of DKG participants.
In our meeting I had suggested to also remove this assumption. However, I am not sure anymore that this is such a good idea, because it might require changes specifically on the crypto layer.
If our implementation becomes super messy with convention 2. in place, I think that would be solid reason to remove this convention. But I think it could be very well possible that this is not the case.
In my head, the biggest source of potential confusion is that we refer to the "DKG-indices" as indices -- and we also have the notion of indices in the consensus committee (based on canonical ordering)
maybe it would help to change our terminology and call them "DKG-identifiers", which are non-negative integers.
we could retain the convention that  "DKG-identifiers" of n DKG participants always cover the interval 0,...,n-1.
In other words, I think calling them something different than indices might already alleviate lots of possible confusion ... while the convention being the numbers 0,...,n-1 could be treated as an implementation detail.